### PR TITLE
Fix error handling in X509_chain_up_ref

### LINKED ISSUE #8592: X509_chain_up_ref does not check the return value of X509_up_ref possibly resulting in malformed stack
https://github.com/openssl/openssl/blob/3e4b43b9e5aba6261b1ceb832c597ce8225782ef/crypto/x509/x509_cmp.c#L448-L458

There is no check of the return value of X509_up_ref, possibly resulting in a malformed chain if a lock acquire fails to occur. Found through code inspection.

### PR DESCRIPTION
**Pull request by mattcaswell**
Fixes #8592

### CODE REVIEW

File: crypto/x509/x509_cmp.c [hunk 1]
```diff
@@ -452,7 +452,16 @@ STACK_OF(X509) *X509_chain_up_ref(STACK_OF(X509) *chain)
...
+        if (!X509_up_ref(x)) {
+            int j;
+
+            for (j = 0; j < i; j++) {
+                x = sk_X509_value(ret, j);
+                X509_free(x);
+            }
+            sk_X509_free(ret);
+            return NULL;
+        }
```
Comment (kroeckx) [inline crypto/x509/x509_cmp.c]: I guess you can make this much simpler like this:
```
ret = sk_X509_new(X509_cmp);
for (i = 0; i < sk_X509_num(chain); i++) {
    X509 *x = sk_X509_value(chain, i);
    if (!X509_up_ref(x)) {
        sk_X509_pop_free(ret, X509_free);
        return NULL;
    }
    sk_X509_push(ret, x);
}
return ret;
```

Comment (kroeckx) [reply crypto/x509/x509_cmp.c]: The sk_X509_new() should probably get the compare function from chain, but we currently don't seem to have a way to get that with public functions.

Comment (mattcaswell) [reply crypto/x509/x509_cmp.c]: Thanks...that gave me another idea.

**Comment by mattcaswell** [comment]
Closing this in favour of #8602.

**Pull request closed without merging**